File: src/afp/identity.py
from packageurl import PackageURL

SUPPORTED_SCHEMES = ("pkg", "https", "http", "mcp", "afp")


class InvalidSubjectUri(ValueError):
    """El subject_uri no respeta ningún esquema soportado por AFP."""


def _scheme_of(uri: str) -> str:
    # pkg: y afp: usan ':'; http(s):// y mcp:// usan '://'
    if "://" in uri:
        return uri.split("://", 1)[0]
    if ":" in uri:
        return uri.split(":", 1)[0]
    return ""


def validate_subject_uri(uri: str) -> str:
    if not uri or not isinstance(uri, str):
        raise InvalidSubjectUri("subject_uri vacío")
    scheme = _scheme_of(uri)
    if scheme not in SUPPORTED_SCHEMES:
        raise InvalidSubjectUri(f"esquema no soportado: {scheme!r}")
    if scheme == "pkg":
        try:
            PackageURL.from_string(uri)
        except ValueError as exc:
            raise InvalidSubjectUri(f"PURL inválido: {exc}") from exc
    else:
        # para http(s)/mcp/afp exigimos algo después del separador
        rest = uri.split("://", 1)[-1] if "://" in uri else uri.split(":", 1)[-1]
        if not rest:
            raise InvalidSubjectUri(f"localizador vacío para {scheme!r}")
    return uri

File: ts/src/identity.ts
/** subject_uri validation (SPEC §5). */
import { PackageURL } from "packageurl-js";

import { InvalidSubjectUri } from "./errors.js";

export const SUPPORTED_SCHEMES = ["pkg", "https", "http", "mcp", "afp"] as const;

export function schemeOf(uri: string): string {
  const authority = uri.indexOf("://");
  if (authority !== -1) return uri.slice(0, authority);
  const colon = uri.indexOf(":");
  if (colon !== -1) return uri.slice(0, colon);
  return "";
}

export function validateSubjectUri(uri: unknown): string {
  if (typeof uri !== "string" || uri.length === 0) {
    throw new InvalidSubjectUri("empty subject_uri");
  }
  const scheme = schemeOf(uri);
  if (!(SUPPORTED_SCHEMES as readonly string[]).includes(scheme)) {
    throw new InvalidSubjectUri(`unsupported scheme: ${JSON.stringify(scheme)}`);
  }
  if (scheme === "pkg") {
    try {
      PackageURL.fromString(uri);
    } catch (exc) {
      throw new InvalidSubjectUri(`invalid PURL: ${(exc as Error).message}`);
    }
    return uri;
  }
  // http(s)/mcp/afp: the locator after the separator must be non-empty.
  const sep = uri.includes("://") ? "://" : ":";
  const rest = uri.slice(uri.indexOf(sep) + sep.length);
  if (rest.length === 0) {
    throw new InvalidSubjectUri(`empty locator for ${JSON.stringify(scheme)}`);
  }
  return uri;
}
